pootle/apps/pootle_store/models.py

```diff
@@ -1854,146 +1724,84 @@ def update(self, overwrite=False, store=None, only_newer=False, user=None,
-                            changes['added'] += 1
-                    else:
-                        update_unitids[uid] = {'dbid': old_unitids[uid]['dbid'],
-                                               'index': new_unit_index}
-
-            # Step 2: mark obsolete units as such
-
-            obsolete_dbids = [old_unitids[uid]['dbid']
-                              for uid in old_unitid_set -
-                                         old_obsolete_unitid_set -
-                                         new_unitid_set]
-            changes['obsolete'] = \
-                self.mark_units_obsolete(obsolete_dbids,
-                                         revision=update_revision)
-
-
-            # Step 3: update existing units
-
-            update_dbids = set([x['dbid'] for x in update_unitids.values()])
-            common_dbids.update(update_dbids)
-
-            # Optimization: only go through unchanged units since the last sync
-            if not overwrite:
-                common_dbids = self.remove_modified_units(common_dbids)
-
-            (changes['updated'],
-             changes['suggested']) = self.update_units(store, common_dbids,
-                                                       update_unitids, user,
-                                                       revision=update_revision,
-                                                       submission_type=submission_type)
-            self.file_mtime = disk_mtime
-
-            if (filter(lambda x: changes[x] > 0, changes) and
-                store == self.file.store):
-                if self.last_sync_revision is not None:
-                    changes['unsynced'] = \
-                        self.increment_unsynced_unit_revision(update_revision)
-
-                self.last_sync_revision = update_revision
-
+                    if len(v) > 0:
+                        change = True
+                        break
+            if change:
+                changes = self.update_from_diff(store,
+                                                store_revision,
+                                                to_change, Revision.incr(),
+                                                user, submission_type,
+                                                resolve_conflict)
         finally:
             # Unlock store
-            self.state = old_state
-            self.save()
-            if filter(lambda x: changes[x] > 0, changes):
-                log(u"[update] %s units in %s [revision: %d]" % (
-                    get_change_str(changes), self.pootle_path,
-                    self.get_max_unit_revision())
-                )
+            if old_state < PARSED:
+                self.state = PARSED
+            else:
+                self.state = old_state
+            has_changed = any(x > 0 for x in changes.values())
+            self.save(update_cache=has_changed)
+            if has_changed:
+                logger.info(
```

Comment:
Note the `log()` function serves its own purpose and logs to the `action` named logger. This change will break existing logging configurations. Since introducing logging-related changes in this PR is not relevant, I'd appreciate taking such changes out and keeping logging as it was.